src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -1943,6 +1937,55 @@ MultipleStackTracesCollector::allocate_and_fill_stacks(jint thread_count) {
          "the last copied frame info must be the last record");
 }
 
+// Supports platform and virtual threads.
+// JvmtiVTMSTransitionDisabler is always set by this function.
+void
+JvmtiHandshake::execute(JvmtiUnitedHandshakeClosure* hs_cl, jthread target) {
+  JavaThread* current = JavaThread::current();
+  HandleMark hm(current);
+
+  JvmtiVTMSTransitionDisabler disabler(target);
+  ThreadsListHandle tlh(current);
+  JavaThread* java_thread = nullptr;
+  oop thread_obj = nullptr;
+
+  jvmtiError err = JvmtiEnvBase::get_threadOop_and_JavaThread(tlh.list(), target, &java_thread, &thread_obj);
+  if (err != JVMTI_ERROR_NONE) {
+    hs_cl->set_result(err);
+    return;
+  }
+  Handle target_h(current, thread_obj);
+  execute(hs_cl, &tlh, java_thread, target_h);
+}
+
+// Supports platform and virtual threads.
+// A virtual thread is always identified by the target_h oop handle.
+// The target_jt is always nullptr for an unmounted virtual thread.
+// JvmtiVTMSTransitionDisabler has to be set before call to this function.
+void
+JvmtiHandshake::execute(JvmtiUnitedHandshakeClosure* hs_cl, ThreadsListHandle* tlh,
+                        JavaThread* target_jt, Handle target_h) {
+  bool self = target_jt == JavaThread::current();
+
+  hs_cl->set_self(self);           // needed when suspend is required for non-current target thread
+  hs_cl->set_target_h(target_h);   // need this to differentiate between virtual and carrier thread
+
+  if (java_lang_VirtualThread::is_instance(target_h())) { // virtual thread
+    if (!JvmtiEnvBase::is_vthread_alive(target_h())) {
```

Comment:
There is only one issue I see in how this check is implemented and the removal of the VM_op for unmounted vthreads. The change of state to TERMINATED happens after notifyJvmtiUnmount(), i.e we can see that this vthread is alive here but a check later can return is not. This might hit the assert in JvmtiEnvBase::get_vthread_jvf() (maybe this the issue you saw on your first prototype). We can either change that order at the Java level, or maybe better change this function to read the state and add a case where if the state is RUNNING check whether the continuation is done or not (jdk_internal_vm_Continuation::done(cont)).

Reply:
Thank you for the suggestion. Will check it.

Reply:
I've added the check for `!jdk_internal_vm_Continuation::done(cont)` into ` JvmtiEnvBase::is_vthread_alive(oop vt)` but then decided to remove it again. This is racy for `JvmtiHandshake` execution. As you correctly stated, the change of state to `TERMINATED` happens after `notifyJvmtiUnmount()`. The target virtual thread will be blocked in the `notifyJvmtiUnmount()` because the `JvmtiVTMSTransitionDisabler` is set. This gives us a guaranty that the target virtual thread won't change its state to `TERMINATED` while a handshake is executed. But it becomes not true if we add the `!jdk_internal_vm_Continuation::done(cont)` check.
Form the other hand, absence of this check allows for target virtual thread stack to become empty (with no frames). This is a known problem but I'd prefer to attack it separately.

Reply:
So the problematic case I'm thinking is when the JvmtiVTMSTransitionDisabler starts after the vthread executed notifyJvmtiUnmount(), i.e the vthread is already outside the transition, but before changing the state to TERMINATED. JvmtiVTMSTransitionDisabler will proceed, and since the carrierThread field has already been cleared we will treat it as an unmounted vthread. Then we can see first that is alive in JvmtiHandshake::execute() but then we could hit the assert that is already TERMINATED in JvmtiEnvBase::get_vthread_jvf().

Reply:
Right, this is a valid concern. Will try to fix this.

Reply:
I'm suggesting to fix it (for the unmounted case only) this way:
```
@@ -1976,6 +1976,13 @@ JvmtiHandshake::execute(JvmtiUnitedHandshakeClosure* hs_cl, ThreadsListHandle* t
       return;
     }
     if (target_jt == nullptr) {    // unmounted virtual thread
+      // JvmtiVTMSTransitionDisabler can start after the vthread executed notifyJvmtiUnmount(), i.e.
+      // the vthread is already outside the transition, but before changing the state to TERMINATED.
+      // Changing the state to TERMINATED is racy, so we check if the continuation is done in advance.
+      oop cont = java_lang_VirtualThread::continuation(target_h());
+      if (jdk_internal_vm_Continuation::done(cont)) {
+        return;
+      }
       hs_cl->do_vthread(target_h); // execute handshake closure callback on current thread directly
     }
   }
```

Reply:
Sounds good. Is there a reason why not have the check inside JvmtiEnvBase::is_vthread_alive()?

Reply:
If it is a part of the `JvmtiEnvBase::is_vthread_alive()` then it is racy for mounted virtual threads.
It is not racy for unmounted virtual threads.

Reply:
So we should only see that a continuation is done for an unmounted vthread. The last place where we could see a mounted vthread is at notifyJvmtiEnd(), blocked in start_VTMS_transition(), but the continuation is not marked done yet. 
Also I realize the window for the problematic case I mentioned starts even earlier at notifyJvmtiEnd(), not notifyJvmtiUnmount(), because blocking due to JvmtiVTMSTransitionDisabler happens in start_VTMS_transition() not finish_VTMS_transition(). Once the vthread executed notifyJvmtiEnd() any JvmtiVTMSTransitionDisabler that happens afterwards will fall into this case. So maybe the first sentence of the comment should instead be: "The vthread could have already executed the last unmount but might not have changed state to TERMINATED yet."

Reply:
Thank you. Now I see that `done` is set to `true` during an unmount transition.
So, I'm convinced to move the check for `jdk_internal_vm_Continuation::done(cont)` to `JvmtiEnvBase::is_vthread_alive()`.